File: threadsafety/part_one.cc
#include <iostream>
#include <mutex>
#include <thread>
#include <vector>

using namespace std;

int counter = 0;

void incrementCounter() {
    for (int i = 0; i < 1000000; ++i) {
        ++counter;
    }
}

int main() {
    vector<thread> threads;
    for (int i = 0; i < 10; i++) {
        threads.push_back(thread(incrementCounter));
    }

    for (int i = 0; i < 10; i++) {
        threads[i].join();
    }

    cout << "Counter = " << counter << endl;
    return 0;
}

File: threadsafety/part_one.rs
use std::thread::scope;

fn main() {
    let mut counter = 0;
    // scoped threads will auto join when the scope ends
    scope(|s| {
        for _ in 0..10 {
            s.spawn(|| {
                for _ in 0..1000000 {
                    counter += 1;
                }
            });
        }
    });
    println!("Counter = {counter}")
}
